File: 0103-binary-tree-zigzag-level-order-traversal/0103-binary-tree-zigzag-level-order-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<List<Integer>> zigzagLevelOrder(TreeNode root) {
        List<List<Integer>> ans = new ArrayList();
        Queue<TreeNode> q = new LinkedList();
        q.add(root);
        int dir=0;
        while(!q.isEmpty()){
            int n = q.size();
            List<Integer> list = new ArrayList();
            for(int i=0; i<n; i++){
                TreeNode node = q.poll();
                if(node==null)continue;
                list.add(node.val);
                if(node.left != null)q.add(node.left);
                if(node.right != null)q.add(node.right);               
            }
            if(dir%2==0){
                if(!list.isEmpty())ans.add(list);            
            }
            else{
                Collections.reverse(list);
                if(!list.isEmpty())ans.add(list); 
            }
            dir++;
        }
        return ans;
    }
}

File: 0103-binary-tree-zigzag-level-order-traversal/0103-binary-tree-zigzag-level-order-traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def zigzagLevelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        if not root: return []
        ans = []
        q = deque([root])
        dire = True
        while q:
            _no = len(q)
            _nodes = []
            for _ in range(_no):
                if dire:
                    node = q.popleft()
                else:
                    node = q.pop()
                _nodes.append(node.val)
                if dire:
                    if node.left:
                        q.append(node.left)
                    if node.right:
                        q.append(node.right)
                else:
                    if node.right:
                        q.appendleft(node.right)
                    if node.left:
                        q.appendleft(node.left)

            dire = not dire
            ans.append(_nodes)
        return ans 
        